File: PreviousVerisons/LetterFactoryV1.0/src/main/java/letterfactory/FileNamer.java
package letterfactory;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class FileNamer {
    private FileNamer() {}

    public static String sanitize(String s) {
        if (s == null) return "";
        // Windows illegal: \ / : * ? " < > |
        String sanitized = s.replaceAll("[\\\\/:*?\"<>|]", "_");
        sanitized = sanitized.replaceAll("_+", "_");
        return sanitized.replaceAll("^_+|_+$", "").trim();
    }

    public static String buildNameFromParts(List<String> placeholderParts, Map<String, String> values) {
        if (placeholderParts == null || placeholderParts.isEmpty()) return "";
        StringBuilder sb = new StringBuilder();
        for (String p : placeholderParts) {
            sb.append(sanitize(values.getOrDefault(p, "")));
        }
        return sb.toString();
    }

    public static String ensureUnique(String base, Set<String> used, int indexWidth) {
        String b = (base == null || base.isBlank()) ? "Document" : base;
        String name = b;
        int i = 1;
        while (used.contains(name)) {
            name = b + "_" + String.format("%0" + indexWidth + "d", i++);
        }
        used.add(name);
        return name;
    }

    public static List<String> normalizeSelectedParts(List<String> selected, List<String> templateOrder) {
        if (selected == null || selected.isEmpty()) return List.of();
        if (templateOrder == null || templateOrder.isEmpty()) return new ArrayList<>(selected);
        var out = new ArrayList<String>();
        for (String p : templateOrder) {
            if (selected.contains(p)) out.add(p);
        }
        return out;
    }
}

File: LetterFactoryV2.2/app/src/filenamer.rs
//! Filename helpers. Port of the Java `FileNamer`.

use std::collections::HashSet;
use std::sync::OnceLock;

use regex::Regex;

fn illegal_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    // Windows illegal: \ / : * ? " < > |
    RE.get_or_init(|| Regex::new(r#"[\\/:*?"<>|]"#).unwrap())
}

fn multi_underscore_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"_+").unwrap())
}

/// Replace filesystem-illegal characters, collapse runs of `_`, and trim
/// leading/trailing underscores and whitespace.
pub fn sanitize(s: &str) -> String {
    let replaced = illegal_re().replace_all(s, "_");
    let collapsed = multi_underscore_re().replace_all(&replaced, "_");
    collapsed.trim_matches('_').trim().to_string()
}

/// Concatenate the sanitized values of the selected placeholder parts.
pub fn build_name_from_parts(parts: &[String], values: &std::collections::HashMap<String, String>) -> String {
    if parts.is_empty() {
        return String::new();
    }
    let mut sb = String::new();
    for p in parts {
        let v = values.get(p).map(|s| s.as_str()).unwrap_or("");
        sb.push_str(&sanitize(v));
    }
    sb
}

/// Ensure the base name is unique within `used`, appending a zero-padded index
/// of width `index_width` on collision. Inserts the chosen name into `used`.
pub fn ensure_unique(base: &str, used: &mut HashSet<String>, index_width: usize) -> String {
    let b = if base.trim().is_empty() { "Document" } else { base };
    let mut name = b.to_string();
    let mut i = 1usize;
    while used.contains(&name) {
        name = format!("{b}_{i:0width$}", width = index_width);
        i += 1;
    }
    used.insert(name.clone());
    name
}
